server/src/main/java/com/hedera/block/server/verification/hasher/ConcurrentStreamingTreeHasher.java

```diff
@@ -0,0 +1,260 @@
+     * empty leaves to reach a power of two), given the penultimate status of the tree and the hash of the last
+     * leaf added to the tree.
+     *
+     * @param penultimateStatus the penultimate status of the tree
+     * @param lastLeafHash the last leaf hash added to the tree
+     * @return the root hash of the tree
+     */
+    public static Bytes rootHashFrom(@NonNull final Status penultimateStatus, @NonNull final Bytes lastLeafHash) {
+        requireNonNull(lastLeafHash);
+        var hash = lastLeafHash.toByteArray();
+        final var rootHeight = rootHeightFor(penultimateStatus.numLeaves() + 1);
+        for (int i = 0; i < rootHeight; i++) {
+            final var rightmostHash = penultimateStatus.rightmostHashes().get(i);
+            if (rightmostHash.length() == 0) {
+                hash = HashingUtilities.combine(hash, HashCombiner.EMPTY_HASHES[i]);
+            } else {
+                hash = HashingUtilities.combine(rightmostHash.toByteArray(), hash);
+            }
+        }
+        return Bytes.wrap(hash);
+    }
+
+    private class HashCombiner {
+        private static final ThreadLocal<MessageDigest> DIGESTS =
+                ThreadLocal.withInitial(HashingUtilities::sha384DigestOrThrow);
+        private static final int MAX_DEPTH = 24;
+        private static final int MIN_TO_SCHEDULE = 16;
+
+        private static final byte[][] EMPTY_HASHES = new byte[MAX_DEPTH][];
+
+        static {
+            EMPTY_HASHES[0] = noThrowSha384HashOf(new byte[0]);
+            for (int i = 1; i < MAX_DEPTH; i++) {
+                EMPTY_HASHES[i] = HashingUtilities.combine(EMPTY_HASHES[i - 1], EMPTY_HASHES[i - 1]);
+            }
+        }
+
+        private final int height;
+
+        private HashCombiner delegate;
+        private List<byte[]> pendingHashes = new ArrayList<>();
+        private CompletableFuture<Void> combination = CompletableFuture.completedFuture(null);
+
+        private HashCombiner(final int height) {
+            if (height >= MAX_DEPTH) {
+                throw new IllegalArgumentException("Cannot combine hashes at height " + height);
+            }
+            this.height = height;
+        }
+
+        public void combine(@NonNull final byte[] hash) {
+            pendingHashes.add(hash);
+            if (pendingHashes.size() == hashCombineBatchSize) {
+                schedulePendingWork();
+            }
+        }
+
+        public CompletableFuture<Bytes> finalCombination() {
+            if (height == rootHeight) {
+                final var rootHash = pendingHashes.isEmpty() ? EMPTY_HASHES[0] : pendingHashes.getFirst();
+                return CompletableFuture.completedFuture(Bytes.wrap(rootHash));
+            } else {
+                if (!pendingHashes.isEmpty()) {
+                    schedulePendingWork();
+                }
+                return combination.thenCompose(ignore -> delegate.finalCombination());
```

Comment:
This might read better using method handle syntax:
```suggestion
                return combination.thenCompose(delegate::finalCombination);
```

Reply:
I tried but due to generics, needs some parsing and then end up being even less readable as is currently. imo we should leave as is.

Reply:
Odd, there should be no parsing difference between a trivial state-capturing lambda (which allocates) and a method handle (which does not allocate).

Ideally we will go through and be fairly diligent about reducing the use of lambdas, but that's a task for future work.